app/controllers/drug_stocks_controller.rb

```diff
@@ -0,0 +1,75 @@
+# This controller is meant to be used from the Android app _only_,
+# hence we handle authentication ourselves from params passed from the client.
+class DrugStocksController < ApplicationController
+  skip_before_action :verify_authenticity_token
+  before_action :authenticate
+  before_action :find_current_facility
+  before_action :set_for_end_of_month
+  layout false
+
+  def create
+    DrugStock.transaction do
+      safe_params[:drug_stocks].map do |drug_stock|
+        DrugStock.create!(facility: current_facility,
+                          user: current_user,
+                          protocol_drug_id: drug_stock[:protocol_drug_id],
+                          received: drug_stock[:received].presence,
+                          in_stock: drug_stock[:in_stock].presence,
+                          for_end_of_month: @for_end_of_month)
+      end
+    end
+    redirect_to drug_stocks_url(for_end_of_month: @for_end_of_month,
+                                facility_id: current_facility.id,
+                                user_id: current_user.id,
+                                access_token: current_user.access_token)
+  rescue ActiveRecord::RecordInvalid => e
+    logger.error "could not create DrugStocks - record invalid", errors: e.message
+    render json: {status: "invalid", errors: e.message}, status: 422
```

Comment:
How to handle this is an open question, I'll add it to the story and ask the android team about it.